documentation: architecture (#13)
kinvolk/seccompagent: docs/architecture.md

---
title: Architecture
weight: 20
description: >
  Architecture of Seccomp Agent.
---

## From Kubernetes to Seccomp Agent

![From Kubernetes to Seccomp Agent](kubernetes-to-seccomp-agent.png "From Kubernetes to Seccomp Agent")

> Steps:

Reviewer: I'd mention that in this example we use containerd, but this _can_ be run with other container runtimes (docker supports it now, for example).

  ```
  {
    "linux": {
      "seccomp": {
        "listenerPath": "/run/seccomp-agent.socket",
        "defaultAction": "SCMP_ACT_ALLOW",
        "syscalls": [
          {
            "names": ["mount"],
            "action": "SCMP_ACT_NOTIFY"
          }
  ```
- Containerd starts runc with `config.json`.
- Runc gets the seccomp file descriptor by calling the seccomp() system call
  via libseccomp-golang.
- Runc sends the seccomp file descriptor to the seccomp agent, along with
  following
  [Container Process State](https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#containerprocessstate):
  ```
  {
    "ociVersion": "0.2.0",
    "fds": ["seccompFd"],
    "pid": 4422,
    "metadata": "",
    "state": {
      "ociVersion": "0.2.0",
      "id": "8eea9f6d...",
      "status": "creating",
      "pid": 4422,
      "bundle": "rootfs",
      "annotations": {}
    }
  }
  ```
- At this point, the Seccomp Agent has the seccomp file descriptor and the
  Container Process State contains the following
  [annotations](https://github.com/containerd/containerd/blob/v1.5.2/pkg/cri/annotations/annotations.go)

>   to identify the pod. Additional information such as the ServiceAccount could
>   be queried to the Kubernetes API.

Reviewer: Why the service account? It seems like a weird example to give. What about just removin that part?

```suggestion
to identify the pod. Additional information can be queried to the Kubernetes API.
```

Author: I can remove it.

I used the ServiceAccount because we could imagine that someone would use the seccomp agent to implement some security service that grants or denies actions based on the ServiceAccount (Pod Security Policy were linked to ServiceAccount too).

  - Pod namespace: `io.kubernetes.cri.sandbox-namespace`
  - Pod name: `io.kubernetes.cri.sandbox-name`
  - Container name: `io.kubernetes.cri.container-name`

## Example of system call performed on behalf of the process

> ![Mount syscall with Seccomp Notify](example-syscall-mount.png "Mount syscall with Seccomp Notify")

Reviewer: Can we give more context (like unprivileged builds)? This is something a user almost never does, it is not clear why we are talking about this, IMHO

Steps:
- A process executes the mount system call:
  ```
  mount("proc", "/proc", "proc", 0, NULL)
  ```
- The seccomp subssytem in the kernel executes the BPF program for the
  seccomp filter.
  The BPF program receives a `struct seccomp_data` as argument, containing:
  ```
  struct seccomp_data {
    int nr;   // syscall mount = 165
    u32 arch; // x86_64
    u64 instruction_pointer;
    u64 args[6]; // syscall args
  }
  ```
- The BPF program returns `SCMP_ACT_NOTIFY`, meaning the decision is deferred
  to a seccomp agent.
- The seccomp agent receives a request from seccomp by running the

>   `SECCOMP_IOCTL_NOTIF_RECV` ioctl on the seccomp file descriptor. The request

Reviewer: The text is clear. But I find it confusing that in the picture the arrow goes from the kernel to the agent. Like, the kernel never calls the agent, but the syscall from the agent just resturns something.

Maybe have the arrow go from the agent to the kernel with the ioctl details, and then an arrow back with the `struct seccomp_notif` can be clearer?

Author: I changed it a bit

  is a `struct seccomp_notif` with the following content:
  ```
  struct seccomp_notif {
    __u64 id;
    __u32 pid;
    __u32 flags;
    struct seccomp_data data;
  };
  ```
- The seccomp agent reads the process' memory to get the arguments of the
  system call.
- The seccomp agent decides to perform the request mount on behalf of the
  process. It enters the mount namespace of the process and calls `mount`.
- The seccomp agent gives a response to the seccomp subsystem by running the
  `SECCOMP_IOCTL_NOTIF_SEND` ioctl on the seccomp file descriptor. The response
  is a `struct seccomp_notif_resp` with the following content:
  ```
  struct seccomp_notif_resp {
    __u64 id;
    __s64 val;   // returns 0
    __s32 error; // errno 0
    __u32 flags;
  };
  ```
- The seccomp subsystem completes the mount system call from the process by
  returning 0 (success) as instructed by the seccomp agent.

> ## Details on kernel synchronisation

Reviewer: Note to self: review this section later

Reviewer: Didn't check the structs, but on a LGTM on cursory look
